src/core/lib/iomgr/event_engine_shims/endpoint.cc

```diff
@@ -0,0 +1,341 @@
+        return true;
+      }
+    }
+  }
+
+  // Decrement the shutdown ref. If this is the last shutdown ref, it also
+  // deletes the underlying event engine endpoint. Deletion of the event
+  // engine endpoint should trigger execution of any pending read/write
+  // callbacks with NOT-OK status.
+  void ShutdownUnref() {
+    if (shutdown_ref_.fetch_sub(1, std::memory_order_acq_rel) ==
+        kShutdownBit + 1) {
+      OnShutdownInternal();
+    }
+  }
+
+  // If trigger shutdown is called the first time, it sets the shutdown bit
+  // and decrements the shutdown ref. If trigger shutdown has been called
+  // before or in parallel, only one of them would win the race. The other
+  // invocation would simply return.
+  void TriggerShutdown() {
+    int64_t curr = shutdown_ref_.load(std::memory_order_acquire);
+    while (true) {
+      if (curr & kShutdownBit) {
+        return;
+      }
+      if (shutdown_ref_.compare_exchange_strong(curr, curr | kShutdownBit,
+                                                std::memory_order_acq_rel,
+                                                std::memory_order_relaxed)) {
+        Ref();
+        if (shutdown_ref_.fetch_sub(1, std::memory_order_acq_rel) ==
+            kShutdownBit + 1) {
+          OnShutdownInternal();
+        }
+        return;
+      }
+    }
+  }
+
+ private:
+  void OnShutdownInternal() {
+    {
+      grpc_core::MutexLock lock(&mu_);
+      fd_ = -1;
+      local_address_ = "";
+      peer_address_ = "";
+    }
+    endpoint_.reset();
+    // For the Ref taken in TriggerShutdown
+    Unref();
+  }
+  std::unique_ptr<EventEngine::Endpoint> endpoint_;
+  std::unique_ptr<grpc_event_engine_endpoint> eeep_;
+  std::atomic<int64_t> refs_{1};
+  std::atomic<int64_t> shutdown_ref_{1};
+  grpc_core::Mutex mu_;
+  std::string peer_address_;
+  std::string local_address_;
+  int fd_{-1};
+};
+
+// Read from the endpoint and place the data in slices slice buffer. The
+// provided closure is also invoked asynchronously.
+void EndpointRead(grpc_endpoint* ep, grpc_slice_buffer* slices,
+                  grpc_closure* cb, bool /* urgent */, int min_progress_size) {
+  auto* eeep =
+      reinterpret_cast<EventEngineEndpointWrapper::grpc_event_engine_endpoint*>(
+          ep);
+  if (!eeep->wrapper->ShutdownRef()) {
+    // Shutdown has already been triggered on the endpoint.
+    grpc_core::ExecCtx::Run(DEBUG_LOCATION, cb, absl::CancelledError());
+    return;
+  }
+
+  eeep->wrapper->Ref();
+  EventEngine::Endpoint::ReadArgs read_args = {min_progress_size};
+  SliceBuffer* read_buffer = new (&eeep->read_buffer) SliceBuffer(slices);
+  eeep->wrapper->Read(
+      [eeep, cb, slices](absl::Status status) {
+        auto* read_buffer = reinterpret_cast<SliceBuffer*>(&eeep->read_buffer);
+        grpc_slice_buffer_move_into(read_buffer->c_slice_buffer(), slices);
+        read_buffer->~SliceBuffer();
+        {
+          grpc_core::ExecCtx exec_ctx;
```

Comment:
Do we need an ApplicationCallbackExecCtx?

Reply:
When run on all the tests it didn't seem like this was needed. I think we can add it back if we encounter some bug.

Reply:
would recommend adding it back... we're probably not testing thoroughly enough, and that callback could trigger *anything*